services/src/main/java/org/openlearn/web/rest/PortfolioItemResource.java

```diff
@@ -144,57 +171,110 @@ public ResponseEntity delete(@PathVariable final Long id) {
-	public ResponseEntity getUploads(@PathVariable final Long portfolioId) {
+	public ResponseEntity getUploads(@PathVariable final Long portfolioId, @ApiParam Pageable pageable) {
 		log.debug("GET request to get course uploads for portfolio " + portfolioId);
-		List<S3ObjectSummary> response = storageService.getUploads(null, portfolioId).getObjectSummaries();
-		return ResponseEntity.ok(response);
+		PortfolioItemDTO portfolioItem = portfolioItemService.findOne(portfolioId);
+		if (portfolioItem == null) throw new PortfolioItemNotFoundException(portfolioId);
+
+		if (canUploadFilesToPortfolio(portfolioItem)) {
+			Page<FileInformationDTO> response = fileInformationService.findAllForPorfolioItem(portfolioId, pageable);
+			return ResponseEntity.ok(response.getContent());
+		} else {
+			return new ResponseEntity(HttpStatus.FORBIDDEN);
+		}
 	}
 
-	@GetMapping(path="/{portfolioId}/upload/{keyName:.+}")
+	@GetMapping(path="/{portfolioId}/upload/{id}")
 	@Secured({AuthoritiesConstants.ADMIN, AuthoritiesConstants.ORG_ADMIN, AuthoritiesConstants.INSTRUCTOR, AuthoritiesConstants.STUDENT})
 	public ResponseEntity getUpload(@PathVariable final Long portfolioId,
-									@PathVariable final String keyName) {
-		log.debug("GET request to get course upload : {}", keyName);
-		InputStream response = storageService.getUpload(null, portfolioId, keyName);
-		if (response != null) {
-			try {
-				byte[] out = org.apache.commons.io.IOUtils.toByteArray(response);
+									@PathVariable final Long id) {
+		log.debug("GET request to get portfolio upload : {}", id);
+		PortfolioItemDTO portfolioItem = portfolioItemService.findOne(portfolioId);
+		FileInformationDTO fileInformationDTO = fileInformationService.findOne(id);
 
-				HttpHeaders responseHeaders = new HttpHeaders();
-				responseHeaders.add("content-disposition", "attachment; filename=" + keyName);
-				// responseHeaders.add("Content-Type", type);
+		if (portfolioItem == null) throw new PortfolioItemNotFoundException(portfolioId);
+		if (fileInformationDTO == null) throw new FileInformationNotFoundException(id);
 
-				return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
-			} catch (Exception e) {
-				new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
+		if (canUploadFilesToPortfolio(portfolioItem) && fileInformationDTO.getPortfolioItemId() == portfolioId) {
+			String fileName = fileInformationService.getFileNameFor(id);
+			InputStream response = storageService.getUpload(id);
+			if (response != null) {
+				try {
+					byte[] out = org.apache.commons.io.IOUtils.toByteArray(response);
+
+					HttpHeaders responseHeaders = new HttpHeaders();
+					responseHeaders.add("content-disposition", "attachment; filename=" + fileName);
+					// responseHeaders.add("Content-Type", type);
+
+					return new ResponseEntity(out, responseHeaders, HttpStatus.OK);
+				} catch (Exception e) {
+					new ResponseEntity("File Not Found", HttpStatus.NOT_FOUND);
```

Comment:
This isn't doing anything.

Reply:
Also a failure in marshalling the file should result in a 500 rather than a 404.